File: pages/refactorings/replace-mutable-fields-with-lenses/02-rename-tree/After.scala
// Same, with the single varName optic applied at every node of the
// tree: how to reach the name is defined once, and the walk only
// decides where it applies.
object After:
  case class Var(name: String, ref: Int)
  enum Expr:
    case EVar(v: Var)
    case EApp(f: Expr, x: Expr)
    case ELam(bind: String, body: Expr)

  import Expr.*

  case class Lens[S, A](view: S => A, set: (S, A) => S):
    def modify(f: A => A): S => S = s => set(s, f(view(s)))
  case class Prism[S, A](preview: S => Option[A], review: A => S):
    def modify(f: A => A): S => S =
      s => preview(s).map(a => review(f(a))).getOrElse(s)
  case class PartialLens[S, A](preview: S => Option[A],
                               modify: (A => A) => S => S)

  def compose[S, M, A](p: Prism[S, M],
                       l: Lens[M, A]): PartialLens[S, A] =
    PartialLens(
      preview = s => p.preview(s).map(l.view),
      modify  = f => s => p.preview(s) match
        case Some(m) => p.review(l.modify(f)(m))
        case None    => s,
    )

  def over[S, A](pl: PartialLens[S, A], f: A => A): S => S =
    pl.modify(f)

  val varP: Prism[Expr, Var] =
    Prism[Expr, Var](
      { case EVar(v) => Some(v); case _ => None },
      EVar(_),
    )
  val nameL: Lens[Var, String] =
    Lens[Var, String](_.name, (v, n) => v.copy(name = n))
  val varName: PartialLens[Expr, String] = compose(varP, nameL)

  // Bottom-up: apply the rewrite at every node, descending first.
  def everywhere(f: Expr => Expr)(e: Expr): Expr = e match
    case EVar(_)     => f(e)
    case EApp(a, b)  => f(EApp(everywhere(f)(a), everywhere(f)(b)))
    case ELam(b, bd) => f(ELam(b, everywhere(f)(bd)))

  def renameAll(e: Expr): Expr =
    everywhere(over(varName, _.toUpperCase))(e)

File: pages/refactorings/replace-mutable-fields-with-lenses/02-rename-tree/After.hs
-- Same, with the single varName optic applied at every node of the
-- tree: how to reach the name is defined once, and the walk only
-- decides where it applies.
module After where

import Data.Char (toUpper)

data Var  = Var  { vName :: String, vRef :: Int }
  deriving (Eq, Show)
data Expr = EVar Var | EApp Expr Expr | ELam String Expr
  deriving (Eq, Show)

data Lens s a = Lens { view :: s -> a, set :: (s, a) -> s }

modifyL :: Lens s a -> (a -> a) -> s -> s
modifyL l f s = set l (s, f (view l s))

data Prism s a = Prism
  { preview :: s -> Maybe a
  , review  :: a -> s
  }

data PartialLens s a = PartialLens
  { plPreview :: s -> Maybe a
  , plModify  :: (a -> a) -> s -> s
  }

composeO :: Prism s m -> Lens m a -> PartialLens s a
composeO p l = PartialLens
  { plPreview = \s -> view l <$> preview p s
  , plModify  = \f s -> case preview p s of
      Nothing -> s
      Just m  -> review p (modifyL l f m)
  }

over :: PartialLens s a -> (a -> a) -> s -> s
over pl = plModify pl

varP :: Prism Expr Var
varP = Prism
  { preview = \e -> case e of EVar v -> Just v; _ -> Nothing
  , review  = EVar
  }

nameL :: Lens Var String
nameL = Lens { view = vName, set = \(v, n) -> v { vName = n } }

varName :: PartialLens Expr String
varName = composeO varP nameL

-- Bottom-up: apply the rewrite at every node, descending first.
everywhere :: (Expr -> Expr) -> Expr -> Expr
everywhere f e@(EVar _)   = f e
everywhere f (EApp a b)   = f (EApp (everywhere f a) (everywhere f b))
everywhere f (ELam b bd)  = f (ELam b (everywhere f bd))

renameAll :: Expr -> Expr
renameAll = everywhere (over varName (map toUpper))
